### PR TITLE
Integrate Docker and changes to Encryptor

### PR DESCRIPTION
**Pull request by jbagnara**
This pull request's purpose primarily serves to both integrate docker into the project for automating setting up the development environment on the jetson nano, and to reimplement multiprocessing into the encryption process.

Among these changes are smaller changes including bug fixes associated with threading, and improving some preexisting documentation

### CODE REVIEW

File: src/jetson/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+pycuda
+pillow
+tqdm==4.42.1
+pycocotools
+pycrypto
+pbkdf2
+distro
+salt
+numpy
```
Comment (stephend017) [inline src/jetson/requirements.txt:11]: evreything in here needs a version you can get them by running `pip freeze` . after doing this validate that your code still works

File: src/jetson/main.py [hunk 2]
```diff
@@ -297,28 +304,28 @@ def encryptFrame(img:np.ndarray,
     g.eval()
     class_names = ['Glasses', 'Goggles', 'Neither']
 
-    cap = VideoCapturer() #Instantiate Video Capturer object
+    cap = VideoCapturer() 
     detector = FaceDetector(detector=args.detector, cuda=args.cuda and torch.cuda.is_available(), set_default_dev=True) #Instantiate Face Detector object
-    cl = Classifier(g) #Instantiate Classifier object
+    cl = Classifier(g)
+    enc = Encryptor()
```
Comment (stephend017) [inline src/jetson/main.py:310]: can we make these variable names more descriptive? `cap` , `c1` and `enc` are hard to read

File: src/jetson/main.py [hunk 3]
```diff
@@ -227,54 +225,62 @@ def classifyFrame(self,
...
-        cv2.imwrite(face_file_name, img)
-    except KeyboardInterrupt:
-        pass
+        if args.write_imgs:
+            if not os.path.isdir(args.output_dir):
+                os.mkdir(args.output_dir)
+            global fileCount
+            face_file_name = os.path.join(args.output_dir, f'{fileCount.value}.jpg')
+    
+            #TODO: Remove this print statement after db integration
```
Comment (stephend017) [inline src/jetson/main.py:279]: can we please address this. either resolve or notify the person in charge of resolving this

File: src/jetson/main.py [hunk 4]
```diff
@@ -121,6 +124,7 @@ def detect(self,
                 bboxes.append((xmin, ymin, xmax, ymax))
             return bboxes
 
+
 class VideoCapturer(object):
```
Comment (stephend017) [inline src/jetson/main.py:128]: please include a docstring here

File: src/jetson/main.py [hunk 5]
```diff
@@ -144,9 +145,9 @@ def update(self):
 
     def get_frame(self):
         ''' Return current frame in video stream'''
+        _, self.frame = self.capture.read()
         return self.frame
 
-
 class Classifier:
```
Comment (stephend017) [inline src/jetson/main.py:151]: please include a docstring here

File: src/jetson/main.py [hunk 6]
```diff
@@ -297,28 +304,28 @@ def encryptFrame(img:np.ndarray,
...
     class_names = ['Glasses', 'Goggles', 'Neither']
 
-    cap = VideoCapturer() #Instantiate Video Capturer object
+    cap = VideoCapturer() 
     detector = FaceDetector(detector=args.detector, cuda=args.cuda and torch.cuda.is_available(), set_default_dev=True) #Instantiate Face Detector object
-    cl = Classifier(g) #Instantiate Classifier object
+    cl = Classifier(g)
+    enc = Encryptor() 
 
     while True:
```
Comment (stephend017) [inline src/jetson/main.py:312]: this needs a comment. Never write `while True` without an explicit comment of its need. 

on a side note, `while True` is bad. it hard to read and all it does is raise alarm flags for me as a developer. I'd much rather see something like 

```python
run_face_detection: bool = True
while run_face_detection:
    # execute code here
    if stop_condition:
        run_face_detection = False
```

I know its more code but its makes it more clear what is happening here. readability is usually better than conciseness when writing code

File: src/jetson/main.py [hunk 7]
```diff
@@ -227,54 +225,62 @@ def classifyFrame(self,
...
+            face_file_name = os.path.join(args.output_dir, f'{fileCount.value}.jpg')
+    
+            #TODO: Remove this print statement after db integration
+            print("writing ", face_file_name)
+            with fileCount.get_lock():
+                fileCount.value += 1
+            cv2.imwrite(face_file_name, img)
 
 
 if __name__ == "__main__":
```
Comment (stephend017) [inline src/jetson/main.py:286]: please see [this comment](https://github.com/PurdueCAM2Project/Embedded2/pull/8#pullrequestreview-437686539) for a good example of how to structure scripts/ main files and try and apply that here. its very messy and somewhat hard to follow at what is going on here

File: src/jetson/main.py (hunk 3 above)
Comment (jbagnara) [reply src/jetson/main.py:279]: Could you clarify what needs to be resolved? db integration has not yet been implemented

File: src/jetson/main.py [hunk 8]
```diff
@@ -265,20 +276,69 @@ def encryptFrame(self, img:np.ndarray,
...
+        boxes: facial Coordinates
+        output_dir: directory to be written to
+    ''' 
+    encryptedImg = encryptor.encryptFrame(img, boxes)
+    if write_imgs:
+        writtenImg = writeImg(encryptedImg, output_dir)
+        encryptRet.put(writtenImg) 
+   
+
+def drawFrame(boxes, frame, fps):
```
Comment (stephend017) [inline src/jetson/main.py:322]: this method needs a docstring

File: src/jetson/main.py (hunk 3 above)
Comment (stephend017) [reply src/jetson/main.py]: Is db integration being worked on then? Sorry I'm not at the meetings so I am a bit clueless to what is being worked on at the moment. If db integration does not exist can you please provide a comment to how this code behaves without db integration and why its needed. Sometimes when features like this get pushed to the side then the person who wrote this comment might not be around to answer why they wrote it if it becomes ambiguous to why it was written in the first place.

As a good rule of thumb for comments. don't say what the code is doing or what the code needs to do in the future, say why the code is doing what its doing. for example.

```python
# TODO: add db integration here 
```

is a lot more ambiguous than 

```python
# TODO: add db integration here so we can store the images collected for future analysis
```

See how the second comment says what the reason for adding db integration was? it makes it a lot easier to understand and potentially saves a meeting in the future by simply giving more context to this comment.

Comment (jbagnara) [reply src/jetson/main.py]: Thanks for the feedback. db integration into main.py is planned after we can get it cleaned up and implement processing changes. We don't intend to replace any of the code by this TODO, just remove the print statement: it only exists now for testing purposes

File: src/jetson/main.py (hunk 4 above)
Comment (jbagnara) [reply src/jetson/main.py]: docstring containing VideoCapturer class information is located in __init__ function. Should this be moved? This is the case for all classes in this script.

**Comment by jbagnara** [comment]
> > a few documentation / stylistic changes need to be made on the code side. as for docker I would like
> > some form of validation that it is in a working state before approving
> 
> The documentation and stylistic changes were good just a few more and it should be good to go. You also haven't addressed my second item about validation that docker is setup correctly and works. This is by far the more important item on that list. I personally don't like to just trust someone that it works since they'll always say "it worked when I ran it" or "it worked on my machine" so validation here is crucial.
> 
> I'm also going to add a 3rd item and ask that you provide some documentation in the form of a `.md` file of how to run this with docker. I'm guessing not everyone here knows how to do that and I know for a fact that most of CAM2 does not know how to do that so this documentation would be good not just for this team but all of CAM2.

I have attached the docker build log to this comment as a form of validation that the image builds and is usable. We are currently waiting on a physical hardware dependency for our kit to arrive and will be unable to run the project itself for validation. However, no dependency issues are identified during preprocessing. Of course, it's still possible we may still run into version specific errors when we are able to test the full detector.

[buildLog.txt](https://github.com/PurdueCAM2Project/Embedded2/files/4847490/buildLog.txt)
